fffw/wrapper/base.py

```diff
@@ -29,30 +50,115 @@ def __post_init__(self) -> None:
+        args = [ensure_binary(self.command)] + self.get_args()
+        return subprocess.Popen(args,
+                                stdin=self.stdin,
+                                stderr=self.stdout,
+                                stdout=self.stderr)
+
+    def handle_stdout_event(self) -> bool:
+        if not self._proc.stdout:
+            return False
+        line = self._proc.stdout.readline()
+        self._output.write(self.handle_stdout(ensure_text(line)))
+        return bool(line)
+
+    def handle_stderr_event(self) -> bool:
+        if not self._proc.stderr:
+            return False
+        line = self._proc.stderr.readline()
+        self._errors.write(self.handle_stderr(ensure_text(line)))
+        return bool(line)
+
+    # noinspection PyAttributeOutsideInit
+    def run(self, timeout: Optional[float] = None) -> Tuple[int, str, str]:
+        self._proc = self.start_process()
+        self._output = io.StringIO()
+        self._errors = io.StringIO()
+        self._timeout = timeout
+        self._deadline = timeout and time.time() + timeout
+        try:
+            with self._proc:
+                handlers, poll = self.init_stream_handers()
+                spin = 1024  # ms
+                while self._proc.poll() is None:
+                    spin = self.poll_process(handlers, poll, spin)
+                for handler in handlers.values():
+                    # read data from buffered streams
+                    while handler():
+                        pass
+
+        finally:
+            if self._proc.returncode is None:
+                self.logger.warning("killing %s", self.command)
+                self._proc.kill()
+            return_code = self._proc.returncode
+            output = self._output.getvalue()
+            errors = self._errors.getvalue()
+            del self._proc
+            del self._output
+            del self._errors
+            del self._timeout
+            del self._deadline
+
+        self.logger.info("%s return code is %s", self.command, return_code)
+        return return_code, output, errors
+
+    def poll_process(self, handlers: Dict[int, Callable[[], bool]],
+                     poll: PollProtocol, spin: int) -> int:
+        for fd, event in poll.poll(spin):
+            handlers[fd]()
+            # speedup stdout/stderr reading if present
+            spin = max(1, spin // 2)
+        else:
+            # slow down if no output is read
+            spin = min(1024, spin * 2)
+            if (self._timeout and self._deadline and
+                    self._deadline < time.time()):
+                self.logger.error("Process %s timeouted",
+                                  self.command)
+                raise subprocess.TimeoutExpired(self._proc.args,
+                                                self._timeout)
+        return spin
+
+    def init_stream_handers(self) -> Tuple[Dict[int, Callable[[], bool]],
+                                           PollProtocol]:
```

Comment:
init_stream_handlers вроде как интерфейс, но при этом перед вызвом нужно выставить self.**_**proc.stdout и self.**_**proc.stderr. Как то не логично. Нельзя ли эти потоки передать параметрами? И назначать self._proc, если они заданы.

Reply:
There are multiple problems with interfaces and typing here:

1. First, wrapper is `@dataclass`. This means that all executable parameters must be passed to `__init__`, because passing it to `run()` makes `@dataclass` pointless.
2. Re-entrance of `run()`. I'd like to guarantee that any wrapper instance could be run multiple times without messing with side-effects, but I don't know how. 
3. `Popen` duck typing depends on it's `__init__` arguments. I.e. passing `stdout=PIPE` makes `Popen.stdout` to be `TextIO`. This leads to multiple unnecessary none-checks.
4. Complexity of `run` method. It makes lot's of stuff and thus is split to multiple pieces.

Looks like some more time is needed to make a better design here.

Reply:
Well... asyncio simplifies everything :)